pkg/messageFormat/lexer.go

```diff
@@ -0,0 +1,129 @@
+package messageFormat
+
+import (
+	"fmt"
+	"strings"
+)
+
+type Token struct {
+	Value string
+	Level int
+	Type  tokenType
+}
+
+type tokenType int
+
+const (
+	Keyword          tokenType = iota
+	PlaceholderOpen  tokenType = iota
+	PlaceholderClose tokenType = iota
+	Literal          tokenType = iota
+	Text             tokenType = iota
+	Function
+	Variable
+)
+
+func Lex(str string) ([]Token, error) {
+	var (
+		tokens           []Token
+		parsedTokens     []Token
+		buf              []rune
+		placeholderLevel int
+	)
+
+	for i, r := range str {
```

Comment:
Note that index won't always represent the position of the character in string, this might only work for ASCII characters. In Go Strings are UTF-8 encoded, this means each charcter called rune can be of 1 to 4 bytes long.

for pos, char := range "日本語" {
    fmt.Printf("character %c starts at byte position %d\n", char, pos)
}
This prints :
character 日 starts at byte position 0
character 本 starts at byte position 3
character 語 starts at byte position 6